flink-core/src/main/java/org/apache/flink/api/connector/source/lib/DataGeneratorSource.java

```diff
@@ -0,0 +1,211 @@
+ * manner, because, despite the fact that the produced stream is bounded, the end bound is pretty
+ * far away.
+ */
+@Experimental
+public class DataGeneratorSource<OUT>
+        implements Source<OUT, NumberSequenceSplit, Collection<NumberSequenceSplit>>,
+                ResultTypeQueryable<OUT> {
+
+    private static final long serialVersionUID = 1L;
+
+    private final SourceReaderFactory<OUT, NumberSequenceSplit> sourceReaderFactory;
+    private final TypeInformation<OUT> typeInfo;
+
+    private final NumberSequenceSource numberSource;
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}.
+     *
+     * @param generatorFunction The {@code GeneratorFunction} function.
+     * @param count The number of generated data points.
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            GeneratorFunction<Long, OUT> generatorFunction,
+            long count,
+            TypeInformation<OUT> typeInfo) {
+        this(generatorFunction, count, -1, typeInfo);
+    }
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}.
+     *
+     * @param generatorFunction The {@code GeneratorFunction} function.
+     * @param count The number of generated data points.
+     * @param sourceRatePerSecond The overall source rate per second (across all source subtasks).
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            GeneratorFunction<Long, OUT> generatorFunction,
+            long count,
+            double sourceRatePerSecond,
+            TypeInformation<OUT> typeInfo) {
+        this(
+                new GeneratorSourceReaderFactory<>(generatorFunction, sourceRatePerSecond),
+                count,
+                typeInfo);
+        ClosureCleaner.clean(
+                generatorFunction, ExecutionConfig.ClosureCleanerLevel.RECURSIVE, true);
+    }
+
+    /**
+     * Instantiates a new {@code DataGeneratorSource}. This constructor allows users can take
+     * control of the low-level data generation details by supplying a custom {@code
+     * SourceReaderFactory}. The instantiated {@code SourceReader}s are expected to produce data
+     * based on processing {@code NumberSequenceSplit}s.
+     *
+     * @param sourceReaderFactory The {@link SourceReader} factory.
+     * @param count The number of generated data points.
+     * @param typeInfo The type of the produced data points.
+     */
+    public DataGeneratorSource(
+            SourceReaderFactory<OUT, NumberSequenceSplit> sourceReaderFactory,
+            long count,
+            TypeInformation<OUT> typeInfo) {
+        this.sourceReaderFactory = checkNotNull(sourceReaderFactory);
+        this.typeInfo = checkNotNull(typeInfo);
+        this.numberSource = new NumberSequenceSource(0, count - 1);
+    }
+
+    /** @return The number of records produced by this source. */
+    public long getCount() {
```

Comment:
Why is this public?

Reply:
To be honest, I just followed a similar layout/fields exposure that Stephan applied in the [NumberSequenceSource], (https://github.com/apache/flink/blob/ac87127cc8ff095f7390096853ba99aee0c7194e/flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java#L87-L93) anticipating that it could be useful for some test suites that could potentially use the `DataGeneratorSource` internally, similar to [OperatorEventSendingCheckpointITCase](https://github.com/apache/flink/blob/d4d0e4a15fa25f9cdcc6667d08b7591a8999a3fb/flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java#L336-L337)

Reply:
Anyone using the source should know ahead of time how many records it produces, because the user supplied that count after all.
The linked it case is a good example of this; it actually got the `from`/`to` via the constructor :shrug: 

The reason I'm raising this is that I'd like to avoid additional methods that reinforce that the source is bounded or that it uses the `NumberSequenceSource` internally.

Reply:
> The linked it case is a good example of this; it actually got the from/to via the constructor shrug
Yeah, I just noticed that when I took a closer look :sweat_smile: 
Removed it.